src/main/scala/org/constellation/storage/RateLimiting.scala

```diff
@@ -0,0 +1,64 @@
+package org.constellation.storage
+
+import cats.effect.concurrent.Ref
+import cats.effect.Sync
+import cats.implicits._
+import io.chrisdavenport.log4cats.Logger
+import org.constellation.primitives.{Schema, Transaction}
+
+import scala.math.ceil
+
+class RateLimiting[F[_]: Sync: Logger]() {
+  private[storage] val counter: Ref[F, Map[Schema.Address, Int]] = Ref.unsafe(Map())
+  private[storage] val blacklisted: StorageService[F, Int] = new StorageService()
+
+  def update(txs: List[Transaction]): F[Unit] =
+    for {
+      grouped <- txs.groupBy(_.src).mapValues(_.size).pure[F]
+      _ <- counter.update(_ |+| grouped)
+
+      _ <- counter.get.flatTap(c => Logger[F].debug(s"Update [rate-limiting]: $c"))
+      _ <- blacklist()
+    } yield ()
+
+  def reset(cbHashes: List[String])(checkpointService: CheckpointService[F]): F[Unit] =
+    for {
+      cbs <- cbHashes.traverse(checkpointService.fullData).map(_.flatten)
+      txs = cbs.flatMap(_.checkpointBlock.get.transactions.toList)
+      grouped = txs.groupBy(_.src).mapValues(_.size)
+      _ <- counter.set(grouped)
```

Comment:
What if here I will update counter while accepting checkpoint, will this set override that update ?

Reply:
Yes. I'll add here a single lock for both reset and update